examples/gno.land/r/system/names/names.gno

```diff
@@ -2,59 +2,197 @@
 package names
 
 import (
+	"errors"
+	"regexp"
 	"std"
 
 	"gno.land/p/demo/avl"
+	"gno.land/p/demo/ufmt"
 )
 
 // "AddPkg" will check if r/system/names exists. If yes, it will
 // inspect the realm's state and use the following variable to
 // determine if an address can publish a package or not.
 var namespaces avl.MutTree // name(string) -> Space
 
+// TODO: more accurate.
+var reNamespace = regexp.MustCompile(`^[a-z][a-z0-9_]{2,30}$`)
+
 type Space struct {
 	Admins  []std.Address
 	Editors []std.Address
 	InPause bool
 }
 
+func (s *Space) isAdmin(addr std.Address) bool {
+	if containsAddress(s.Admins, addr) {
+		return true
+	}
+	return false
+}
+
+func (s *Space) addAdmin(newAdmin std.Address) {
+	if !containsAddress(s.Admins, newAdmin) {
+		s.Admins = append(s.Admins, newAdmin)
+	}
+}
+
+func (s *Space) removeAdmin(admin std.Address) error {
+	if len(s.Admins) == 1 {
+		return errors.New("namespace at least needs one admin")
+	}
+	admins := removeAddress(s.Admins, admin)
+	s.Admins = admins
+	return nil
+}
+
+func (s *Space) addEditor(newEditor std.Address) {
+	if !containsAddress(s.Editors, newEditor) {
+		s.Editors = append(s.Editors, newEditor)
+	}
+}
+
+func (s *Space) removeEditor(editor std.Address) error {
+	editors := removeAddress(s.Editors, editor)
+	s.Editors = editors
+	return nil
+}
+
+func (s *Space) hasPerm(caller std.Address) bool {
+	if s.InPause {
+		return false
+	}
+
+	if containsAddress(s.Admins, caller) {
+		return true
+	}
+
+	if containsAddress(s.Editors, caller) {
+		return true
+	}
+
+	return false
+}
+
 func Register(namespace string) {
 	// TODO: input sanitization:
 	// - already exists / reserved.
 	// - min/max length, format.
 	// - fees (dynamic, based on length).
-	panic("not implemented")
+	if existsNamespace(namespace) {
+		panic("namespace already exists")
+	}
+
+	if err := validateNamespace(namespace); err != nil {
+		panic(err)
+	}
+
+	caller := std.GetOrigCaller()
+	namespaces.Set(namespace, &Space{
+		Admins: []std.Address{caller},
+	})
 }
 
 func AddAdmin(namespace string, newAdmin std.Address) {
-	// TODO: assertIsAdmin()
-	panic("not implemented")
+	space := getSpace(namespace)
+	assertIsAdmin(space)
+	space.addAdmin(newAdmin)
 }
 
-func RemoveAdmin(namespace string, newAdmin std.Address) {
-	// TODO: assertIsAdmin()
-	// TODO: check if self.
-	panic("not implemented")
+func RemoveAdmin(namespace string, admin std.Address) {
+	space := getSpace(namespace)
+	assertIsAdmin(space)
+	err := space.removeAdmin(admin)
```

Comment:
```suggestion
        // TODO: prevent self
	err := space.removeAdmin(admin)
```

I suggest we remove the possibility of demoting ourselves.

Reply:
Does that mean the caller can't remove the caller himself?

Reply:
Caller can remove itself by creating another account, promoting it as admin, and the new account to demote the first one.

Reply:
It checks to prevent self removing. 
https://github.com/gnolang/gno/pull/384/commits/dac208f0dc9abe38955af9c4b175db70b7eced37